src/codeu/chat/server/Server.java

```diff
@@ -276,11 +305,80 @@ public void run() {
         final ConversationHeader conversation = view.findConversation(conversationId);
         final Message message = view.findMessage(messageId);
         relay.write(id,
-                    secret,
-                    relay.pack(user.id, user.name, user.creation),
-                    relay.pack(conversation.id, conversation.title, conversation.creation),
-                    relay.pack(message.id, message.content, message.creation));
+                secret,
+                relay.pack(user.id, user.name, user.creation),
+                relay.pack(conversation.id, conversation.title, conversation.creation),
+                relay.pack(message.id, message.content, message.creation));
       }
     };
   }
+
+  private void process(String line) throws IOException {
+    //turns log string into an array of words
+    String[] splitLog = line.split(" ");
+
+    //uuid and time occupy the same spot in each format so they have been extracted ahead of time
+    Uuid uuid = Uuid.parse(splitLog[1]);
+    long time = Long.parseLong(splitLog[3]);
+
+    switch (splitLog[0]) {
+      case "ADD-USER":
+        String name = splitLog[2];
+        controller.newUser(uuid, name, Time.fromMs(time));
+        break;
+      case "ADD-CONVERSATION":
+        String title = splitLog[2];
+        Uuid owner = Uuid.parse(splitLog[4]);
+        controller.newConversation(uuid, title, owner, Time.fromMs(time));
+        break;
+      case "ADD-MESSAGE":
+        String message = splitLog[2];
+        Uuid conversationUuid = Uuid.parse(splitLog[4]);
+        Uuid senderUuid = Uuid.parse(splitLog[5]);
+        controller.newMessage(uuid, senderUuid, conversationUuid, message, Time.fromMs(time));
+        break;
+    }
+  }
+
+  public void restore(File persistentPath) {
+
+    File log = new File(persistentPath.getPath());
+
+    try {
+
+      boolean created = log.createNewFile(); // true if file created, false otherwise
+      //read in and restore state if the file existed
+      if (!created) {
+        String line;
+        BufferedReader reader = new BufferedReader(new FileReader(log));
+        while ((line = reader.readLine()) != null) {
+          process(line);
+        }
+        reader.close();
+      }
+
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  public void checkBuffer() {
+
+    File log = new File(persistentPath.getPath());
+
+    if (logBuffer.size() == 15) {
+      try {
+
+        BufferedWriter writer = new BufferedWriter(new FileWriter(log));
+        for (int i = 0; i < 15; i++) {
+          writer.write(logBuffer.remove());
+          writer.newLine();
+        }
+        writer.close();
+
+      } catch (IOException e) {
+        e.printStackTrace();
+      }
+    }
+  }
```

Comment:
just pass the file path to the controller when we add a new object, and write the transactions in the controller. The server should not worry about logging logic. It should just pass required info to the controller and the controller should do the work. This of this like lazy execution, you don't actually put the logic as long as you can avoid it. this ensures that all your critical logic stays restricted to few files, which makes things very modular and maintainable.